File: src/javascript/redos.js
// CWE-1333: Inefficient Regular Expression Complexity (ReDoS).
// Rule: js/redos, js/polynomial-redos
// Autofix: typically NOT provided — fixing requires redesigning the regex
// itself (often non-backtracking alternative or a length cap), which depends
// on the developer's intent. Use the Copilot Coding Agent.

const express = require('express');
const app = express();

const EMAIL_RE = /^([a-zA-Z0-9]+)+@([a-zA-Z0-9]+)+\.[a-zA-Z]{2,}$/;
const TRIM_RE = /^\s+|\s+$/;
const CSS_COLOR_RE = /^(#([0-9a-fA-F]+)+|rgb\((\s*\d+\s*,)+\s*\d+\s*\))$/;

app.get('/validate-email', (req, res) => {
  const email = String(req.query.email || '');
  res.json({ ok: EMAIL_RE.test(email) });
});

app.post('/trim', express.text(), (req, res) => {
  const body = req.body || '';
  res.send(body.replace(TRIM_RE, ''));
});

app.get('/color', (req, res) => {
  res.json({ ok: CSS_COLOR_RE.test(String(req.query.c || '')) });
});

app.listen(3100);

File: src/python/redos.py
"""CWE-1333 / CWE-730: ReDoS via catastrophic backtracking.

Rule: py/redos, py/polynomial-redos
Autofix: typically NOT provided — fix requires redesigning the regex
(remove nested quantifiers, anchor properly, switch to a parser, or
impose a length cap). Use the Copilot Coding Agent.
"""

import re
from flask import Flask, request

app = Flask(__name__)

EMAIL_RE = re.compile(r"^([a-zA-Z0-9]+)+@([a-zA-Z0-9]+)+\.[a-zA-Z]{2,}$")
CSS_COLOR_RE = re.compile(r"^(#([0-9a-fA-F]+)+|rgb\((\s*\d+\s*,)+\s*\d+\s*\))$")
TRIM_RE = re.compile(r"^\s+|\s+$")


@app.route("/email")
def email():
    e = request.args.get("e", "")
    return {"ok": bool(EMAIL_RE.match(e))}


@app.route("/color")
def color():
    c = request.args.get("c", "")
    return {"ok": bool(CSS_COLOR_RE.match(c))}


@app.route("/trim", methods=["POST"])
def trim():
    return TRIM_RE.sub("", request.data.decode())


if __name__ == "__main__":
    app.run(port=4100)
